src/core/labeling/qgspallabeling.cpp

```diff
@@ -2373,119 +2373,159 @@ std::unique_ptr<QgsLabelFeature> QgsPalLayerSettings::registerFeatureWithDetails
     }
   }
 
-  if ( mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionX ) )
+  bool hasDataDefinedPosition = false;
   {
-    exprVal = mDataDefinedProperties.value( QgsPalLayerSettings::PositionX, context.expressionContext() );
-    if ( !exprVal.isNull() )
+    bool ddPosition = false;
+
+    if ( mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionX )
+         && mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionY )
+         && !mDataDefinedProperties.value( QgsPalLayerSettings::PositionX, context.expressionContext() ).isNull()
+         && !mDataDefinedProperties.value( QgsPalLayerSettings::PositionY, context.expressionContext() ).isNull() )
+    {
+      ddPosition = true;
+
+      bool ddXPos = false, ddYPos = false;
+      xPos = mDataDefinedProperties.value( QgsPalLayerSettings::PositionX, context.expressionContext() ).toDouble( &ddXPos );
+      yPos = mDataDefinedProperties.value( QgsPalLayerSettings::PositionY, context.expressionContext() ).toDouble( &ddYPos );
+      if ( ddXPos && ddYPos )
+        hasDataDefinedPosition = true;
+    }
+    else if ( mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionPoint )
+              && !mDataDefinedProperties.value( QgsPalLayerSettings::PositionPoint, context.expressionContext() ).isNull() )
     {
-      if ( !exprVal.isNull() )
-        xPos = exprVal.toDouble( &ddXPos );
+      ddPosition = true;
 
-      if ( mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionY ) )
+      QVariant pointAsVariant = mDataDefinedProperties.value( QgsPalLayerSettings::PositionPoint, context.expressionContext() );
+      QgsPoint point;
+      if ( pointAsVariant.canConvert<QgsReferencedGeometry>() )
       {
-        exprVal = mDataDefinedProperties.value( QgsPalLayerSettings::PositionY, context.expressionContext() );
-        if ( !exprVal.isNull() )
+        point = QgsPoint( pointAsVariant.value<QgsReferencedGeometry>().asPoint() );
```

Comment:
We probably **should** be transforming the point here, not ignoring the point's CRS entirely :wink:

Reply:
I have introduced a warning if the point crs is not equal to the layer crs. Theoretically it is possible to have a different crs for the additional geometry but i can't see practical cases where it is the case.